### PR TITLE
guardrails: add observability for attribution

### PR DESCRIPTION
**Pull request by keegancsmith**
This adds traces, logs and metrics via the Operation abstraction. It is my first time using it, but I believe I am using it correctly. To use this we introduce a constructor which initializes the Operation for the attribution service.

Test Plan: go test and sg start and visiting the prometheus metrics.

Part of #53766 
Part of #52366

### CODE REVIEW

File: enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go [hunk 1]
```diff
@@ -108,7 +138,10 @@ func (c *Service) SnippetAttribution(ctx context.Context, snippet string, limit
 	return &agg, nil
 }
 
-func (c *Service) snippetAttributionLocal(ctx context.Context, snippet string, limit int) (*SnippetAttributions, error) {
+func (c *Service) snippetAttributionLocal(ctx context.Context, snippet string, limit int) (result *SnippetAttributions, err error) {
+	ctx, traceLogger, endObservation := c.operations.snippetAttributionLocal.With(ctx, &err, observation.Args{})
```
Comment (stefanhengl) [inline enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go:142]: Limit and snippet length are not interesting here?

File: enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go [hunk 2]
```diff
@@ -177,7 +210,10 @@ func (c *Service) snippetAttributionLocal(ctx context.Context, snippet string, l
 	}, nil
 }
 
-func (c *Service) snippetAttributionDotCom(ctx context.Context, snippet string, limit int) (*SnippetAttributions, error) {
+func (c *Service) snippetAttributionDotCom(ctx context.Context, snippet string, limit int) (result *SnippetAttributions, err error) {
+	ctx, traceLogger, endObservation := c.operations.snippetAttributionDotCom.With(ctx, &err, observation.Args{})
```
Comment (stefanhengl) [inline enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go:214]: same as above

**Review by stefanhengl — APPROVED** [review]
LGTM

**Review by camdencheek — APPROVED** [review]

File: enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go (hunk 1 above)
Comment (keegancsmith) [reply enterprise/cmd/frontend/internal/guardrails/attribution/attribution.go:142]: yeah they are the same value in the parent trace.

**Pull request merged**